### PR TITLE
[clang][dataflow] Add `Environment::allows()`.

### PR DESCRIPTION
**Pull request by martinboehme**
This allows querying whether, given the flow condition, a certain formula still
has a solution (though it is not necessarily implied by the flow condition, as
`flowConditionImplies()` would check).

This can be checked today, but only with a double negation, i.e. to check
whether, given the flow condition, a formula F has a solution, you can check
`!Env.flowConditionImplies(Arena.makeNot(F))`. The double negation makes this
hard to reason about, and it would be nicer to have a way of directly checking
this.

For consistency, this patch also renames `flowConditionImplies()` to `proves()`;
the old name is kept around for compatibility but deprecated.

### CODE REVIEW

File: clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h [hunk 1]
```diff
@@ -129,9 +129,17 @@ class DataflowAnalysisContext {
   /// token.
   Atom joinFlowConditions(Atom FirstToken, Atom SecondToken);
 
-  /// Returns true if and only if the constraints of the flow condition
-  /// identified by `Token` imply that `Val` is true.
-  bool flowConditionImplies(Atom Token, const Formula &);
+  /// Returns true if the constraints of the flow condition identified by
+  /// `Token` imply that `F` is true.
+  /// Returns false the flow condition does not imply `F` or if the solver times
```
Comment (sgatev) [inline clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h:134]: Missing "if" - "Returns false the flow"

File: clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h [hunk 2]
```diff
@@ -184,6 +192,12 @@ class DataflowAnalysisContext {
   addTransitiveFlowConditionConstraints(Atom Token,
                                         llvm::SetVector<const Formula *> &Out);
 
+  /// Returns true if the solver is able to prove that there is a satisfying
+  /// assignment for `Constraints`
```
Comment (sgatev) [inline clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h:196]: Missing dot at the end of the sentence.

File: clang/include/clang/Analysis/FlowSensitive/DataflowEnvironment.h [hunk 3]
```diff
@@ -546,12 +546,29 @@ class Environment {
...
   /// Returns true if the formula is always true when this point is reached.
-  /// Returns false if the formula may be false, or if the flow condition isn't
-  /// sufficiently precise to prove that it is true.
-  bool flowConditionImplies(const Formula &) const;
+  /// Returns false if the formula may be false (or the flow condition isn't
+  /// sufficiently precise to prove that it is true) or if the solver times out.
+  ///
+  /// Note that there is an asymmetry between this function and `allows()` in
+  /// that they both return false if the solver times out. The assumption is
+  /// that if `proves()` or `allows() ` returns true, this will result in a
```
Comment (sgatev) [inline clang/include/clang/Analysis/FlowSensitive/DataflowEnvironment.h:560]: Extra space at the end of \`allows() \`.

**Comment by sgatev** [comment]
LGTM

**Review by Xazax-hun — APPROVED** [review]

**Comment by martinboehme** [comment]
CI failures look unrelated

File: clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h (hunk 1 above)
Comment (martinboehme) [reply clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h]: Done.

File: clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h (hunk 2 above)
Comment (martinboehme) [reply clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h]: Done.

File: clang/include/clang/Analysis/FlowSensitive/DataflowEnvironment.h (hunk 3 above)
Comment (martinboehme) [reply clang/include/clang/Analysis/FlowSensitive/DataflowEnvironment.h]: Done.

**Review by sgatev — APPROVED** [review]

**Pull request merged**